"""Segregating E gene and Whole Genomes and performing quaility control

This module takes aligned virus sequences and segregates the envelope gene (EG) and whole genomes (WG). It then performs quality control on the sequences.

Params:
  WG Threshold (number): Whole Genome Threshold
  EG Threshold (number): Envelope Gene Threshold
  Serotypes (list of str): List of serotype names (e.g. Dengue_1, etc.)
"""
configfile: "config/config.yaml"


indir = config["input_namespace"]
outdir = config["output_namespace"]
params = config["params"]


def script(name=""):
    from snakemake.remote import AUTO

    filename = srcdir(f"scripts/{name}")
    try:
        return AUTO.remote(filename)
    except TypeError:
        return filename


rule split_genome_and_QC:
    input:
        fasta_files=expand(
            "results/{indir}/Aligned_{serotype}/nextalign.aligned.fasta",
            indir=indir,
            serotype=params["Serotypes"],
        ),
        script=script("Seperate_EG_and_WG.R"),
    output:
        E_gene_dir=expand(
            "results/{outdir}/{serotype}_EG.fasta",
            outdir=outdir,
            serotype=params["Serotypes"],
        ),
        WG_gene_dir=expand(
            "results/{outdir}/{serotype}_WG.fasta",
            outdir=outdir,
            serotype=params["Serotypes"],
        ),
    params:
        wg_threshold=params["WG Threshold"],
        eg_threshold=params["EG Threshold"],
        indir=f"results/{indir}",
        outdir=f"results/{outdir}",
    log:
        "logs/split_genome_and_QC.log",
    message:
        (
            "Segregating E gene and whole genomes from aligned Dengue virus sequences "
            "and performing quality control."
        )
    conda:
        "envs/conda.yaml"
    shell:
        """
        Rscript {input.script} \
            --WG_threshold {params.wg_threshold} \
            --EG_threshold {params.eg_threshold} \
            --input_dir {params.indir} \
            --output_dir {params.outdir} \
            > {log} 2>&1
        """


rule _test:
    input:
        f"results/{outdir}/Dengue_1_EG.fasta",
        f"results/{outdir}/Dengue_2_EG.fasta",
        f"results/{outdir}/Dengue_3_EG.fasta",
        f"results/{outdir}/Dengue_4_EG.fasta",
